libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,611 @@
+{
+    MQTTStatus_t status = MQTTSuccess;
+    size_t recordIndex = MQTT_STATE_ARRAY_MAX_COUNT;
+    MQTTQoS_t tempQoS = MQTTQoS0;
+    MQTTPublishState_t tempState = MQTTStateNull;
+
+    recordIndex = _findInRecord( pMqttContext->outgoingPublishRecords,
+                                 MQTT_STATE_ARRAY_MAX_COUNT,
+                                 packetId,
+                                 &tempQoS,
+                                 &tempState );
+
+    if( qos == MQTTQoS0 )
+    {
+        status = MQTTSuccess;
+    }
+    else if( packetId == MQTT_PACKET_ID_INVALID )
+    {
+        status = MQTTBadParameter;
+    }
+    /* Make sure there's no collision. */
+    else if( recordIndex >= MQTT_STATE_ARRAY_MAX_COUNT )
+    {
+        status = _addRecord( pMqttContext->outgoingPublishRecords,
+                             MQTT_STATE_ARRAY_MAX_COUNT,
+                             packetId,
+                             qos,
+                             MQTTPublishSend );
+    }
+    else
+    {
+        /* If the ID was found there must be a collision. */
+        status = MQTTStateCollision;
+    }
+    
+    return status;
+}
+
+MQTTPublishState_t MQTT_CalculateStatePublish( MQTTStateOperation_t opType,
+                                               MQTTQoS_t qos )
+{
+    MQTTPublishState_t calculatedState = MQTTStateNull;
+    switch( qos )
+    {
+        case MQTTQoS0:
+            calculatedState = MQTTPublishDone;
+            break;
+        case MQTTQoS1:
+            calculatedState = ( opType == MQTT_SEND )? MQTTPubAckPending : MQTTPubAckSend;
+            break;
+        case MQTTQoS2:
+            calculatedState = ( opType == MQTT_SEND )? MQTTPubRecPending : MQTTPubRecSend;
+            break;
+        default:
+            /* No other QoS values. */
+            break;
+    }
+    return calculatedState;
+}
+
+MQTTPublishState_t MQTT_UpdateStatePublish( MQTTContext_t * pMqttContext,
+                                            uint16_t packetId,
+                                            MQTTStateOperation_t opType,
+                                            MQTTQoS_t qos )
+{
+    MQTTPublishState_t newState = MQTTStateNull;
+    MQTTPublishState_t currentState = MQTTStateNull;
+    MQTTStatus_t mqttStatus = MQTTSuccess;
+    size_t recordIndex = MQTT_STATE_ARRAY_MAX_COUNT;
+    MQTTQoS_t foundQoS = MQTTQoS0;
+    MQTTPubAckInfo_t * records = NULL;
+    bool isTransitionValid = false;
+
+    /* Set this pointer so some common code can be used later. */
+    if( opType == MQTT_SEND )
+    {
+        records = pMqttContext->outgoingPublishRecords;
+    }
+    else
+    {
+        records = pMqttContext->incomingPublishRecords;
+    }
+
+    recordIndex = _findInRecord( records,
+                                 MQTT_STATE_ARRAY_MAX_COUNT,
+                                 packetId,
+                                 &foundQoS,
+                                 &currentState );
+    
+    if( qos == MQTTQoS0 )
```

Comment:
Same comment as above (line 332).  First two conditions could be checked before _findInRecord. The second condition could be an assert.

Reply:
Rearranged so this check happens before `findInRecord`